Unified core

cgwalters opened the pull request:
WIP for https://github.com/projectatomic/rpm-ostree/issues/729

May extract some of the prep to a separate PR soon.

cgwalters commented:
Depends: https://github.com/ostreedev/ostree/pull/1105

cgwalters commented:
OK, this is updated again against master.  Required for [rpm-ostree jigdo ♲📦](https://github.com/projectatomic/rpm-ostree/issues/1081).  Still WIP...going to try to clean things up a bit.

cgwalters commented:
Note to self, need to implement `NODOCS`, that's an immediate obvious diff between libdnf and ucore `compose tree`.  Tentatively thinking of doing this at unpack time.

cgwalters commented:
OK, lifting WIP on this.  I have been squashing issues in the unprivileged path for a while but I am not sure how many more remain, and none of those really block the jigdo work.  So I think my vote is let's land this, it's just an `--ex` option, but already works well enough I think for people to test out.

cgwalters commented on src/app/rpmostree-compose-builtin-tree.c:
```diff
@@ -412,6 +410,35 @@ install_packages_in_root (RpmOstreeTreeComposeContext  *self,
+      if (!glnx_opendirat (AT_FDCWD, "/", TRUE, &host_rootfs_dfd, error))
+        return FALSE;
+      g_autoptr(OstreeSePolicy) sepolicy = ostree_sepolicy_new_at (host_rootfs_dfd, cancellable, error);
+      if (!sepolicy)
```
I should note this also means that `--ex-unified-core` requires having `selinux-policy-targeted` in the container/host.

jlebon replied:
Let's just print a notice instead and keep going? It would make it easier to test from a container. :)
Anyway if caching, then we only pay that relabel cost once, right?

cgwalters replied:
One can install the policy in a container, that's what I've done.  We *could* stumble on but it means the relabel actually rewrites every object.  Seems easier to require people to install policy?

(That said of course this whole thing intersects with https://github.com/projectatomic/rpm-ostree/issues/971)

jlebon commented on src/app/rpmostree-compose-builtin-tree.c:
```diff
@@ -475,52 +502,77 @@ install_packages_in_root (RpmOstreeTreeComposeContext  *self,
+    {
+      if (!rpmostree_context_import (self->corectx, cancellable, error))
+        return FALSE;
+      if (!rpmostree_context_relabel (self->corectx, cancellable, error))
```
We don't need this relabel though, right? Otherwise IIUC, we will unnecessarily relabel the whole cache everytime the build env changes policy, even if the target tree didn't.

cgwalters replied:
It seems you're right...I am not entirely sure what I was thinking there.  Deleted in a fixup :arrow_down:

jlebon commented on tests/compose-tests/libbasic-test.sh:
```diff
@@ -0,0 +1,41 @@
+# This used to live in test-basic.sh, but it's now shared with test-basic-unified.sh
```
Minor: any reason for formatting this as a lib? Seems cleaner as a `basic-test.sh` we just source directly like ostree's `basic-test.sh`. Not super opposed either way though.

cgwalters replied:
I kept this as just to avoid too much churn but may revisit later.

jlebon commented:
Looks like `f26-compose` is failing with:

```
+ runcompose --ex-unified-core --add-metadata-from-json metadata.json
+ rpm-ostree compose tree --repo /var/tmp/rpmostree-compose-cache-0/repo-build --cache-only --cachedir=/var/tmp/rpmostree-compose-cache-0/cache composedata/fedora-basic.json --ex-unified-core --add-metadata-from-json metadata.json
No previous commit for fedora/stable/x86_64/basic
error: With policy root '/proc/self/fd/18/etc/selinux/targeted': selabel_open(SELABEL_CTX_FILE): No such file or directory
```

cgwalters commented:
bot, retest this please

jlebon commented:
Looks good! Though seems like CI is now failing with:

```
+ kver=4.11.8
+ ostree --repo=/var/tmp/rpmostree-compose-cache-0/repo-build ls fedora/stable/x86_64/basic /usr/lib/modules/4.11.8/vmlinuz /usr/lib/modules/4.11.8/initramfs.img
error: No such file or directory: /usr/lib/modules/4.11.8
+ _cleanup_tmpdir
+ test -z ''
+ test -f /var/tmp/rpm-ostree-compose-test.86iBZK/.test
+ rm /var/tmp/rpm-ostree-compose-test.86iBZK -rf
FAIL
```

jlebon commented:
> And even when run as uid 0, there are some bugs, though I'm not sure
of any showstoppers yet.  For example, dracut's `dracut-install` calls
`cp --preserve=xattrs` which fails to copy the `user.ostreemeta` xattrs
from a checkout (it shouldn't be copying that anyways...)

Yup, definitely also hitting this now. Did you investigate why it happens? It looks like we're getting `ENOTSUP`, which is *probably* due to `/tmp` being on tmpfs. But yet AFAIK, we're using `--dir /tmp` when invoking bwrap and our rootfs is in the ostree repo tmpdir, right?

cgwalters commented:
Yeah it's the `user.ostreemeta` xattr, and "no user. on tmpfs" combo.  See 61c761bc2c35fb244d46fbbde97161f5927071dc

cgwalters commented:
(I suppose we could actually bind a host `/var/tmp` tmpdir?)

jlebon commented:
Ahhh right,  bwrap itself uses a tmpfs when creating the container root. Yeah, probably easier to just bind a `/var/tmp` tmpdir and drop the `--tmpdir=/tmp` argument. We can do that later though!

Seems like the unified test is still failing.

cgwalters commented:
I am not sure what's going on, that `kver` is clearly wrong; must be something with the regexp?  But I can't reproduce locally, and it sure seems to me the kernel version downloaded in the logs should be fine too.  I added some more logging in a fixup and also reworked it to be a bit more robust.  (I keep forgetting about `grep -o`)

cgwalters commented:
Well, that fixed it anyways!

jlebon commented:
@rh-atomic-bot r+ cb761dc
